### PR TITLE
Block Hooks: Allow returning null from hooked_block filter

### PR DESCRIPTION
**Pull request by ockham**
The [`hooked_block_types` filter](https://developer.wordpress.org/reference/hooks/hooked_block_types/) allows extenders to conditionally insert a block, based on the anchor block's ''type''.

However, it's currently not possible to base the criterion for insertion on the value of an attribute of the anchor block.

One fairly straight-forward way to implement this would be to allow the `hooked_block` (and `hooked_block_{$hooked_block_type}`) filter -- which has access to the entire parsed anchor block, i.e. including attributes -- to return `null`, which Block Hooks could then interpret as "do not render hooked block". (This would avoid changing the `hooked_block_types` filter's signature in an awkward way.)

Trac ticket: https://core.trac.wordpress.org/ticket/60580

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by swissspidy** [comment]
> This would avoid changing the hooked_block_types filter's signature in an awkward way

You still need to change the signature to indicate the value can be null or an array

**Comment by ockham** [comment]
> > This would avoid changing the hooked_block_types filter's signature in an awkward way
> 
> You still need to change the signature to indicate the value can be null or an array

What I meant is I don't need to add a `$parsed_anchor_block` arg to [`hooked_block_types`](https://developer.wordpress.org/reference/hooks/hooked_block_types/) (when it already has `$anchor_block_type`).

It's true that I need to allow for `hooked_block` and `hooked_block_{$hooked_block_type}` to return `null`, but that's less of a problem -- they've only been introduced during the 6.5 cycle (unlike `hooked_block_types`, which has been around since 6.4).

**Comment by ockham** [comment]
FYI @tjcafferkey -- if you'd like to give this a review 😄

**Comment by swissspidy** [comment]
```
@param array|null $parsed_hooked_block The parsed block array for the given hooked block type or null to remove the block.
```

**Comment by ockham** [comment]
> ```
> @param array|null $parsed_hooked_block The parsed block array for the given hooked block type or null to remove the block.
> ```

@swissspidy Thank you, good point! Added in 556646426d. (I opted to write "to suppress the block" rather than "remove", as the latter would sound a bit like the block has already been inserted and is being removed. LMK if that's okay, or if you'd prefer "remove" after all.)

**Comment by ockham** [comment]
Rebased.

File: tests/phpunit/tests/blocks/setIgnoredHookedBlocksMetadata.php [hunk 1]
```diff
@@ -144,4 +144,36 @@ public function test_set_ignored_hooked_blocks_metadata_for_block_added_by_conte
...
+			if ( 'tests/anchor-block' === $parsed_anchor_block['blockName'] && 'after' === $relative_position ) {
+				return null;
+			}
+
+			return $parsed_hooked_block;
+		};
+
+		add_filter( 'hooked_block_tests/hooked-block-suppressed-by-filter', $filter, 10, 4 );
+		set_ignored_hooked_blocks_metadata( $anchor_block, 'after', $hooked_blocks, null );
+		remove_filter( 'hooked_block_tests/hooked-block-suppressed-by-filter', $filter, 10 );
```
Comment (swissspidy) [inline tests/phpunit/tests/blocks/setIgnoredHookedBlocksMetadata.php]:
```suggestion
		remove_filter( 'hooked_block_tests/hooked-block-suppressed-by-filter', $filter );
```

10 is the default

File: tests/phpunit/tests/blocks/insertHookedBlocks.php [hunk 2]
```diff
@@ -180,4 +180,42 @@ public function test_insert_hooked_blocks_filter_can_wrap_block() {
...
+				return null;
+			}
+
+			return $parsed_hooked_block;
+		};
+		add_filter( 'hooked_block_' . self::HOOKED_BLOCK_TYPE, $filter, 10, 4 );
+		$actual = insert_hooked_blocks( $anchor_block, 'after', self::HOOKED_BLOCKS, array() );
+		remove_filter( 'hooked_block_' . self::HOOKED_BLOCK_TYPE, $filter );
+
+		$this->assertSame( '', $actual, "No markup should've been generatedfor hooked block suppressed by filter." );
```
Comment (swissspidy) [inline tests/phpunit/tests/blocks/insertHookedBlocks.php]:
```suggestion
		$this->assertSame( '', $actual, "No markup should've been generated for hooked block suppressed by filter." );
```

Comment (ockham) [reply tests/phpunit/tests/blocks/insertHookedBlocks.php]: Jinx! 😄 Done in https://github.com/WordPress/wordpress-develop/pull/6143/commits/7360423d23093e4b9234d8db0fa018e78e2bcc52.

**Comment by tjcafferkey** [comment]
> FYI @tjcafferkey -- if you'd like to give this a review 😄

I'm going to try and review this today if required 😄

**Comment by swissspidy** [comment]
@tjcafferkey  If you'd like to, the sooner the better - beta 2 release party kicks off in around 2hrs.

File: src/wp-includes/blocks.php [hunk 3]
```diff
@@ -922,6 +922,10 @@ function insert_hooked_blocks( &$parsed_anchor_block, $relative_position, $hooke
 		 */
 		$parsed_hooked_block = apply_filters( "hooked_block_{$hooked_block_type}", $parsed_hooked_block, $hooked_block_type, $relative_position, $parsed_anchor_block, $context );
 
+		if ( null === $parsed_hooked_block ) {
```
Comment (gziolo) [inline src/wp-includes/blocks.php:925]: What's the reason `hooked_block` filter is not integrated here?

Comment (gziolo) [reply src/wp-includes/blocks.php:925]: Ok, the code snippet was folded 😄

It means that it brings consistency into how these filters work with the ability to return `null`. All good 👍🏻

File: tests/phpunit/tests/blocks/setIgnoredHookedBlocksMetadata.php [hunk 4]
```diff
@@ -144,4 +144,36 @@ public function test_set_ignored_hooked_blocks_metadata_for_block_added_by_conte
 			$anchor_block['attrs']['metadata']['ignoredHookedBlocks']
 		);
 	}
+
+	/**
+	 * @ticket 60580
+	 *
+	 * @covers ::set_ignored_hooked_blocks_metadata
+	 */
+	public function test_set_ignored_hooked_blocks_metadata_for_block_suppressed_by_filter() {
```
Comment (gziolo) [inline tests/phpunit/tests/blocks/setIgnoredHookedBlocksMetadata.php:153]: I had a hard time understanding why the `tests/hooked-block-suppressed-by-filter` block type got removed from the list of ignored hooked blocks. I realized it's related to the filter name, so maybe it should be made more obvious:

```suggestion
	public function test_set_ignored_hooked_blocks_metadata_for_specific_block_type_suppressed_by_filter() {
```

Another idea would be to put the name of the block in a variable and using it in the expression, example:

```php
add_filter( 'hooked_block_' + $block_suppressed_by_filter, $filter, 10, 4 );
```

**Review by gziolo — APPROVED** [review]
Code changes look good to me, but I would appreciate sanity check from @tjcafferkey.

**Review by tjcafferkey — APPROVED** [review]
I've reviewed the code and had a quick catch up with @ockham and from my POV these LGTM. Thanks @ockham!

File: src/wp-includes/blocks.php (hunk 3 above)
Comment (ockham) [reply src/wp-includes/blocks.php:925]: Not sure I'm following; can you rephrase? 😅 

The filter can only modify the `$parsed_hooked_block`; the `$hooked_block_type` is there in case the `$parsed_hooked_block` is e.g. wrapped in a Group block, in which case the top-level `$parsed_hooked_block['blockName']` will change (whereas the `$hooked_block_type` is kept unchanged so that it can be used to set `ignoredHookedBlocks` correctly).

File: tests/phpunit/tests/blocks/setIgnoredHookedBlocksMetadata.php (hunk 4 above)
Comment (ockham) [reply tests/phpunit/tests/blocks/setIgnoredHookedBlocksMetadata.php:153]: Ah, good point. I had an idea for a slightly different test that should be a bit easier to follow; I'll give that a try.

File: src/wp-includes/blocks.php (hunk 3 above)
Comment (gziolo) [reply src/wp-includes/blocks.php:925]: Please ignore it. All good. I responded in https://github.com/WordPress/wordpress-develop/pull/6143#discussion_r1495960855.

File: tests/phpunit/tests/blocks/setIgnoredHookedBlocksMetadata.php (hunk 4 above)
Comment (gziolo) [reply tests/phpunit/tests/blocks/setIgnoredHookedBlocksMetadata.php:153]: It can be a follow-up after Beta 2 :)

**Comment by ockham** [comment]
Committed to Core in https://core.trac.wordpress.org/changeset/57668.

File: tests/phpunit/tests/blocks/setIgnoredHookedBlocksMetadata.php (hunk 4 above)
Comment (ockham) [reply tests/phpunit/tests/blocks/setIgnoredHookedBlocksMetadata.php:153]: Here's a minor rewrite that should make things a bit easier to read:

```diff
diff --git a/tests/phpunit/tests/blocks/setIgnoredHookedBlocksMetadata.php b/tests/phpunit/tests/blocks/setIgnoredHookedBlocksMetadata.php
index d5e9c4e23f..7145e56cde 100644
--- a/tests/phpunit/tests/blocks/setIgnoredHookedBlocksMetadata.php
+++ b/tests/phpunit/tests/blocks/setIgnoredHookedBlocksMetadata.php
@@ -163,16 +163,20 @@ class Tests_Blocks_SetIgnoredHookedBlocksMetadata extends WP_UnitTestCase {
                );
 
                $filter = function ( $parsed_hooked_block, $hooked_block_type, $relative_position, $parsed_anchor_block ) {
-                       if ( 'tests/anchor-block' === $parsed_anchor_block['blockName'] && 'after' === $relative_position ) {
+                       if (
+                               'tests/hooked-block-suppressed-by-filter' === $hooked_block_type &&
+                               'after' === $relative_position &&
+                               'tests/anchor-block' === $parsed_anchor_block['blockName']
+                       ) {
                                return null;
                        }
 
                        return $parsed_hooked_block;
                };
 
-               add_filter( 'hooked_block_tests/hooked-block-suppressed-by-filter', $filter, 10, 4 );
+               add_filter( 'hooked_block', $filter, 10, 4 );
                set_ignored_hooked_blocks_metadata( $anchor_block, 'after', $hooked_blocks, null );
-               remove_filter( 'hooked_block_tests/hooked-block-suppressed-by-filter', $filter );
+               remove_filter( 'hooked_block', $filter );
 
                $this->assertSame( array( 'tests/hooked-block' ), $anchor_block['attrs']['metadata']['ignoredHookedBlocks'] );
        }
```

Comment (gziolo) [reply tests/phpunit/tests/blocks/setIgnoredHookedBlocksMetadata.php:153]: That would help, indeed 💯

Comment (ockham) [reply tests/phpunit/tests/blocks/setIgnoredHookedBlocksMetadata.php:153]: Committed to Core in https://core.trac.wordpress.org/changeset/57676.
